src/Bicep.Core/Semantics/Namespaces/SystemNamespaceType.cs

```diff
@@ -1000,6 +1000,39 @@ maxLength switch
                         return new(LanguageConstants.Object);
                     }, LanguageConstants.Object)
                     .Build();
+
+                yield return new FunctionOverloadBuilder("nameof")
+                    .WithGenericDescription("Returns the name of the specified parameter, variable, or resource.")
+                    .WithRequiredParameter("symbol", LanguageConstants.Any, "The parameter, variable, or resource to retrieve the name of.")
+                    .WithReturnResultBuilder((model, diagnostics, call, argumentTypes) =>
+                    {
+                        var x = call.Arguments[0].Expression switch
+                        {
+                            VariableAccessSyntax variableAccess => variableAccess.Name.IdentifierName,
+                            PropertyAccessSyntax propertyAccess => propertyAccess.PropertyName.IdentifierName,
+                            ResourceAccessSyntax resourceAccess => resourceAccess.ResourceName.IdentifierName,
+                            ModuleDeclarationSyntax moduleDeclaration => moduleDeclaration.Name.IdentifierName,
+                            _ => null,
+                        };
+                        if (x is null)
+                        {
+                            return new(ErrorType.Create(DiagnosticBuilder.ForPosition(call.Arguments[0]).CompileTimeConstantRequired()));
```

Comment:
this is not the exact error that will be raised, I'm just producing something to get glimpse if it's working or not.